Commit hash arg

FoamyGuy opened the pull request:
This adds `--commit-hash` parameter to the update command. It lets you specify a specific commit hash from a build hosted in S3. ex. `python blinka.py update --commit-hash bd87201`

There are still a few things that need to be addressed before this gets merged. Mainly that the post update check is checking version only and thinking that the update failed.

I wanted to go ahead and create this PR to see if you had any thoughts or feedback about this functionality or implementation @bcr 

It uses `boto3` to search the s3 files. I made it a soft requirement by catching the import error. I intend to catch the error in the command and let the user know they must have `boto3` in order to use this flag of the update command. Does that seem like a reasonable approach to you?

bcr commented on urlutil.py:
```diff
@@ -26,3 +31,18 @@ def get_local_file_from_url(url, tempdir):
 
     return pathname
+
+def find_firmware_by_hash(hash, board, language):
```
We're mixing "URL things" with "CircuitPython firmware things" here -- maybe this should be in a different util class.

FoamyGuy replied:
moved to `s3util` in latest commit, lmk if you prefer different name.

bcr commented on urlutil.py:
```diff
@@ -26,3 +31,18 @@ def get_local_file_from_url(url, tempdir):
+
+def find_firmware_by_hash(hash, board, language):
+    resource = boto3.resource('s3')
+    resource.meta.client.meta.events.register('choose-signer.s3.*', disable_signing)
```
Tell me more about this unfortunate bit of code? Is there some reason we can't download the URL directly from the bucket?

FoamyGuy replied:
This was the way I had to do it in order to fetch data from s3 without authenticating the client. I was kind of weary about adding the `boto3` dependency anyway. Perhaps it's possible to skip that and try to pull data with urllib or something else higher level than this s3 wrapper.

bcr commented on urlutil.py:
```diff
@@ -26,3 +31,18 @@ def get_local_file_from_url(url, tempdir):
+    resource.meta.client.meta.events.register('choose-signer.s3.*', disable_signing)
+    bucket = resource.Bucket('adafruit-circuit-python')
+
+    for item in bucket.objects.filter(Prefix="bin/{}/{}/".format(board, language)):
```
So I dorked around with this a bit and I can see the URL https://adafruit-circuit-python.s3.amazonaws.com/index.html?prefix=bin/feather_m0_rfm69/en_US/ has what we need in it. Can we use that instead of Boto3? For some reason the S3 bucket futzing is making me sad.

bcr replied:
I see now that this is a bunch of client JS code and not helpful

bcr commented on commands/update.py:
```diff
@@ -87,7 +88,7 @@ def do_update(args):
             time.sleep(9)
             (version, board_id) = board.identify(args.root)
-            if version == new_version:
+            if version == new_version or (args.commit_hash and new_version in version[-10:]):
```
Is there a simpler expression that will work for both? Like I'm not particularly concerned about "version must exactly equal new_version" so if this is just "if new_version is anywhere in version" I'm OK with that if it works in both cases.

FoamyGuy replied:
I think that should work but not 100% certain. I'll try it out tonight and try to make sure it won't lead to any false positives.

FoamyGuy commented:
I tested out a few scenarios and it does seem to be working fine. I think the biggest risk is weirdness with a beta or rc version somehow tricking it into a false positive. But thats not so bad, and I'm not even sure how big that risk is, just kind of a hunch.

I think this is pretty close to ready to go from my end. The one other tweak I might make visually is to use some fancy emoji instead of "WARNING" in the message that gets printed if user is missing boto3 and they try to use the commit hash arg. I need to poke around a bit and see how you are printing those.

bcr commented on commands/update.py:
```diff
@@ -45,16 +45,27 @@ def do_update(args):
+            url = urlutil.get_s3_url(s3_path)
+            new_version = args.commit_hash
+        else:
+            print("")
```
I think you want something else here. Like some kind of log.critical and an exception thrown. You can probably find another example of this somewhere else, but I think if you hit this we will definitely not be going to space today.

FoamyGuy replied:
Whoops, thank you. critical log added in latest commit. I think it's okay to just print and not throw the exception but lmk if you want it to raise in addition.

bcr commented on urlutil.py:
```diff
@@ -3,6 +3,11 @@
+    import boto3
+    from botocore.handlers import disable_signing
+except ImportError:
+    logging.warn("commit hash param will not work without boto3")
```
No one cares about this unless they are doing a firmware update that needs boto3, so maybe don't output this all the time and just complain about it later when you need it.

FoamyGuy replied:
agree'd for sure. Moved to when you attempt to use the parameter in the newest commit.

FoamyGuy commented:
I changed but then reverted the error log emoji to `{No Entry}` because the pouty face is not red in my terminal. Not sure if that is specific to me or not, but if that commonly not including red I would be in favor of changing to `{No Entry}` or pretty much any emoji that does include red. 

I noticed that the critical one does so I switched to using that.
